File: HouseRobber2.java
import java.util.*;

public class HouseRobber2 {
    public int rob(int[] nums) {
        if (nums.length == 0) {
            return 0;
        } else if (nums.length == 1) {
            return nums[0];
        }
        
        int rob1 = 0;
        int rob2 = 0;

        for (int i = 0; i < nums.length - 1; i++) {
            int temp = Math.max(rob1 + nums[i], rob2);
            rob1 = rob2;
            rob2 = temp;
        }

        int rob3 = 0;
        int rob4 = 0;

        for (int i = 1; i < nums.length; i++) {
            int temp = Math.max(rob3 + nums[i], rob4);
            rob3 = rob4;
            rob4 = temp;
        }
        return Math.max(rob2, rob4);
    }
}

File: HouseRobber2.py
def rob(nums):
    """
    :type nums: List[int]
    :rtype: int
    """
    
    if len(nums) == 1:
        return nums[0]
    if len(nums) == 2:
        return max(nums[0], nums[1])
    
    return max(helper(nums[1:]), helper(nums[:-1]))
    
def helper(nums):
    dp = []
    for i in range(len(nums)):
        dp.append(0)
    dp[0] = nums[0]
    dp[1] = max(nums[0], nums[1])
    
    for i in range(2, len(nums)):
        dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
    
    print(dp)
    return dp[len(nums) - 1]

def main():
    rob([200,3,140,20,10])

if __name__ == "__main__":
    main()